internal/client/client.go

```diff
@@ -297,6 +295,13 @@ func (c *RPCClient) getConnArray(addr string, enableBatch bool, opt ...func(cfg
 			return nil, err
 		}
 	}
+
+	// An idle connArray will not change to active again, this avoid the race condition
+	// that recycling idle connection close an active connection unexpectedly (idle -> active).
+	if array.batchConn != nil && array.isIdle() {
```

Comment:
I confirmed return error will not make TiDB unavailable.
This error will be recognized as a RPC error, and TiDB will refresh the region cache ... and retry the request.

Reply:
> I confirmed return error will not make TiDB unavailable.
> This error will be recognized as a RPC error, and TiDB will refresh the region cache ... and retry the request.

This behavior is a magic,  it's hard to figure out that tidb will retry the request while reading the source code. Could you add some notes here and file an issue? Also, we should do refactor to make sure we know when should return a "retry" error.